### PR TITLE
Refactor/io gene

### PR DESCRIPTION
**Pull request by akaviaLab**
This PR adds typing annotations for cobra.core.gene and  test_gpr. test_gene has only one function, and it was typed. I guess test_gene and test_gpr can be folded into one file, but that seems like a different PR. It also upgrades the code style for them to maintain Python 3.6+ compatibility.

### CODE REVIEW

File: src/cobra/core/gene.py [hunk 1]
```diff
@@ -54,13 +52,37 @@ class GPRWalker(NodeVisitor):
     """
 
     def __init__(self):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
    def __init__(self, **kwargs) -> None:
```

File: src/cobra/core/gene.py [hunk 2]
```diff
@@ -54,13 +52,37 @@ class GPRWalker(NodeVisitor):
...
+        """Initialize a new object.
+
+        Other Parameters
+        ----------------
+        **kwargs:
+            Further keyword arguments are passed on to the parent class.
+
+        """
         NodeVisitor.__init__(self)
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
        super().__init__(**kwargs)
```

File: src/cobra/core/gene.py [hunk 3]
```diff
@@ -74,10 +96,34 @@ class GPRCleaner(NodeTransformer):
     """
 
     def __init__(self):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
    def __init__(self, **kwargs) -> None:
```

File: src/cobra/core/gene.py [hunk 4]
```diff
@@ -74,10 +96,34 @@ class GPRCleaner(NodeTransformer):
...
+        """Initialize a new object.
+
+        Other Parameters
+        ----------------
+        **kwargs:
+            Further keyword arguments are passed on to the parent class.
+
+        """
         NodeTransformer.__init__(self)
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
        super().__init__(**kwargs)
```

File: src/cobra/core/gene.py [hunk 5]
```diff
@@ -87,13 +133,25 @@ def visit_Name(self, node):
         return node
 
     def visit_BinOp(self, node):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]: Missing types here for `node` and return type.

File: src/cobra/core/gene.py [hunk 6]
```diff
@@ -137,7 +195,8 @@ class Gene(Species):
         used.
     """
 
-    def __init__(self, id=None, name="", functional=True):
+    # noinspection PyShadowingBuiltins
+    def __init__(self, id: str = None, name: str = "", functional: bool = True):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
    def __init__(self, id: str = None, name: str = "", functional: bool = True, **kwargs) -> None:
```

File: src/cobra/core/gene.py [hunk 7]
```diff
@@ -154,7 +213,7 @@ def __init__(self, id=None, name="", functional=True):
         self._functional = functional
```
Comment (Midnighter) [inline src/cobra/core/gene.py:213]: Line above should be

```py
super().__init__(id=id, name=name, **kwargs)
```

File: src/cobra/core/gene.py [hunk 8]
```diff
@@ -248,29 +307,23 @@ def remove_from_model(
...
             <tr>
-                <td><strong>Gene identifier</strong></td><td>{id}</td>
+                <td><strong>Gene identifier</strong></td><td>{self.id}</td>
             </tr><tr>
-                <td><strong>Name</strong></td><td>{name}</td>
+                <td><strong>Name</strong></td><td>{self.name}</td>
             </tr><tr>
                 <td><strong>Memory address</strong></td>
-                <td>{address}</td>
+                <td>{"0x0%x" % id(self)}</td>
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
                <td>{id(self):#x}</td>
```

File: src/cobra/core/gene.py [hunk 9]
```diff
@@ -283,8 +336,18 @@ class GPR(Module):
...
+        Parameters
+        ----------
+        gpr_from: Expression, Module, AST
+            An AST expression that will be parsed to GPR.
+        **kwargs:
+            Further keyword arguments are passed on to the parent class.
+        """
         super().__init__(**kwargs)
         self._genes = set()
+        # noinspection PyTypeChecker
```
Comment (Midnighter) [inline src/cobra/core/gene.py]: What's the type error that you see here? Could try
```suggestion
        # noinspection PyTypeChecker
        self.body: Optional[whatever body is] = None
```

File: src/cobra/core/gene.py [hunk 10]
```diff
@@ -533,10 +597,10 @@ def copy(self):
         return deepcopy(self)
 
     def __repr__(self):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
    def __repr__(self) -> str:
```

File: src/cobra/core/gene.py [hunk 11]
```diff
@@ -533,10 +597,10 @@ def copy(self):
...
-            self.__class__.__module__,
-            self.__class__.__qualname__,
-            self.to_string(),
+        """Return the GPR with module, class, and code to recreate it."""
+        return (
+            f"{self.__class__.__module__}.{self.__class__.__qualname__}"
+            f"({self.to_string()!r})"
         )
 
     def __str__(self):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
    def __str__(self) -> str:
```

File: src/cobra/core/gene.py [hunk 12]
```diff
@@ -555,15 +619,11 @@ def __str__(self):
         return self.to_string(names={})
 
     def _repr_html__(self):
```
Comment (Midnighter) [inline src/cobra/core/gene.py]:
```suggestion
    def _repr_html__(self) -> str:
```

Comment (Midnighter) [reply src/cobra/core/gene.py]: Also needs to be `_repr_html_` in order to work with Jupyter notebook.

**Review by Midnighter — CHANGES_REQUESTED** [review]
Just a few additions and some changes to the types.

File: src/cobra/core/gene.py (hunk 7 above)
Comment (akaviaLab) [reply src/cobra/core/gene.py:213]: Species doesn't have **kwargs, so I made it slightly different

File: src/cobra/core/gene.py (hunk 9 above)
Comment (akaviaLab) [reply src/cobra/core/gene.py]: Great suggestion. Thank you. That worked!

**Comment by akaviaLab** [comment]
Closed, since #1167 has the same changes in a cleaner way.

**Pull request closed without merging**